component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -361,24 +366,52 @@ private boolean isValidSignature(String clientId, SignedJWT signedJWT, String te
                                      String alias) throws OAuthClientAuthnException {
 
         X509Certificate cert = null;
+        String jwksUri = "";
+        boolean isValidSignature = false;
         try {
             cert = (X509Certificate) OAuth2Util.getX509CertOfOAuthApp(clientId, tenantDomain);
         } catch (IdentityOAuth2Exception e) {
             if (log.isDebugEnabled()) {
                 log.debug(e.getMessage(), e);
             }
         }
+        //if cert is null check whether a jwks endpoint is configured for the service provider
+        if (cert == null) {
+            try {
+                ServiceProviderProperty[] spProperties = OAuth2Util.getServiceProvider(clientId).getSpProperties();
+                for (ServiceProviderProperty spProperty : spProperties) {
+                    if (Constants.JWKS_URI.equals(spProperty.getName())) {
+                        jwksUri = spProperty.getValue();
+                    }
+                }
+                //validate the signature of the assertion using the jwks end point
+                String jwtString = signedJWT.getParsedString();
```

Comment:
What if cert==null and jwksUri=="" seems logic is not correct.

Reply:
Do we have a way to control the precedence ? For an example one might be using a shared SP and he might need to use jwks for this validation even though a certificate is configured for the SP.

Reply:
For a service provider, either the jwks endpoint or a certificate can be configured.

Reply:
Do we enforce such a thing from UI or service level ? How do we enforce this ?

Reply:
Yes it is enforced from the UI level. A radio button is used to select either the jwks or certificate for SP

Reply:
Can we enforce it from service level as well ?. Otherwise there can be an inconsistant state if someone tries to do the same using service. WDYT ?